### ISSUE TITLE
problem for pre-Training

### ISSUE THREAD

**Issue by carina1754** [Tai-Hsien/MeshSegNet#32]
Hello Tai, 
Currently pretraining is in progress, but the following error occurs. Have you ever seen a case like this? Although overfitting seems to occur, the loss is expressed in nan. If you know, please reply, thank you.

[Epoch: 1/200, Batch: 20/327] loss: nan, dsc: 0.2118112862110138, sen: 0.37746421322226525, ppv: 0.682047963514924
[Epoch: 1/200, Batch: 40/327] loss: nan, dsc: 0.20393260250566528, sen: 0.2627519380301237, ppv: 0.9411806464195251
[Epoch: 1/200, Batch: 60/327] loss: nan, dsc: 0.21566311114002018, sen: 0.27448244094848634, ppv: 0.9411806464195251
[Epoch: 1/200, Batch: 80/327] loss: nan, dsc: 0.2802147898823023, sen: 0.3390341307967901, ppv: 0.9411806464195251
[Epoch: 1/200, Batch: 100/327] loss: nan, dsc: 0.26843334920704365, sen: 0.32725268304347993, ppv: 0.9411806464195251
[Epoch: 1/200, Batch: 120/327] loss: nan, dsc: 0.24501073211431504, sen: 0.30383005663752555, ppv: 0.9411806464195251
[Epoch: 1/200, Batch: 140/327] loss: nan, dsc: 0.27161582112312316, sen: 0.3302940882742405, ppv: 0.9412529110908509
[Epoch: 1/200, Batch: 160/327] loss: nan, dsc: 0.23331434409483337, sen: 0.2921336704865098, ppv: 0.9411806464195251
[Epoch: 1/200, Batch: 180/327] loss: nan, dsc: 0.24789126142859458, sen: 0.3067105896770954, ppv: 0.9411806464195251
[Epoch: 1/200, Batch: 200/327] loss: nan, dsc: 0.3242554787546396, sen: 0.38307480588555337, ppv: 0.9411806464195251
[Epoch: 1/200, Batch: 220/327] loss: nan, dsc: 0.33010296858847143, sen: 0.38892230167984965, ppv: 0.9411806464195251
[Epoch: 1/200, Batch: 240/327] loss: nan, dsc: 0.2743280174618121, sen: 0.33314734827727077, ppv: 0.9411806464195251

**Comment by carina1754** [comment]
I don't know why the loss is nan.  T^T

**Comment by Tai-Hsien** [comment]
Hello,

I would like to clarify first. Did you re-train the pre-trained model and get a nan value in loss?

**Comment by carina1754** [comment]
This is result for y_pred using code like this

def Generalized_Dice_Loss(y_pred, y_true, class_weights, smooth = 1.0):
    '''
    inputs:
        y_pred [n_classes, x, y, z] probability
        y_true [n_classes, x, y, z] one-hot code
        class_weights
        smooth = 1.0
    '''
    smooth = 1.
    loss = 0.
    n_classes = y_pred.shape[-1]
    print(y_pred)    
    for c in range(0, n_classes):
        pred_flat = y_pred[:, :, c].reshape(-1)
        true_flat = y_true[:, :, c].reshape(-1)
        intersection = (pred_flat * true_flat).sum()
       
        # with weight
        w = class_weights[c]/class_weights.sum()
        loss += w*(1 - ((2. * intersection + smooth) /
                         (pred_flat.sum() + true_flat.sum() + smooth)))
        
    return loss

tensor([[[nan, nan, nan,  ..., nan, nan, nan],
         [nan, nan, nan,  ..., nan, nan, nan],
         [nan, nan, nan,  ..., nan, nan, nan],
         ...,
         [nan, nan, nan,  ..., nan, nan, nan],
         [nan, nan, nan,  ..., nan, nan, nan],
         [nan, nan, nan,  ..., nan, nan, nan]],

        [[nan, nan, nan,  ..., nan, nan, nan],
         [nan, nan, nan,  ..., nan, nan, nan],
         [nan, nan, nan,  ..., nan, nan, nan],
         ...,
         [nan, nan, nan,  ..., nan, nan, nan],
         [nan, nan, nan,  ..., nan, nan, nan],
         [nan, nan, nan,  ..., nan, nan, nan]],

        [[nan, nan, nan,  ..., nan, nan, nan],
         [nan, nan, nan,  ..., nan, nan, nan],
         [nan, nan, nan,  ..., nan, nan, nan],
         ...,
         [nan, nan, nan,  ..., nan, nan, nan],
         [nan, nan, nan,  ..., nan, nan, nan],
         [nan, nan, nan,  ..., nan, nan, nan]],

        ...,

        [[nan, nan, nan,  ..., nan, nan, nan],
         [nan, nan, nan,  ..., nan, nan, nan],
         [nan, nan, nan,  ..., nan, nan, nan],
         ...,
         [nan, nan, nan,  ..., nan, nan, nan],
         [nan, nan, nan,  ..., nan, nan, nan],
         [nan, nan, nan,  ..., nan, nan, nan]],

        [[nan, nan, nan,  ..., nan, nan, nan],
         [nan, nan, nan,  ..., nan, nan, nan],
         [nan, nan, nan,  ..., nan, nan, nan],
         ...,
         [nan, nan, nan,  ..., nan, nan, nan],
         [nan, nan, nan,  ..., nan, nan, nan],
         [nan, nan, nan,  ..., nan, nan, nan]],

        [[nan, nan, nan,  ..., nan, nan, nan],
         [nan, nan, nan,  ..., nan, nan, nan],
         [nan, nan, nan,  ..., nan, nan, nan],
         ...,
         [nan, nan, nan,  ..., nan, nan, nan],
         [nan, nan, nan,  ..., nan, nan, nan],
         [nan, nan, nan,  ..., nan, nan, nan]]], device='cuda:0',
       grad_fn=<ViewBackward0>)
tensor([[[nan, nan, nan,  ..., nan, nan, nan],
         [nan, nan, nan,  ..., nan, nan, nan],
         [nan, nan, nan,  ..., nan, nan, nan],
         ...,
         [nan, nan, nan,  ..., nan, nan, nan],
         [nan, nan, nan,  ..., nan, nan, nan],
         [nan, nan, nan,  ..., nan, nan, nan]],

        [[nan, nan, nan,  ..., nan, nan, nan],
         [nan, nan, nan,  ..., nan, nan, nan],
         [nan, nan, nan,  ..., nan, nan, nan],
         ...,
         [nan, nan, nan,  ..., nan, nan, nan],
         [nan, nan, nan,  ..., nan, nan, nan],
         [nan, nan, nan,  ..., nan, nan, nan]],

        [[nan, nan, nan,  ..., nan, nan, nan],
         [nan, nan, nan,  ..., nan, nan, nan],
         [nan, nan, nan,  ..., nan, nan, nan],
         ...,
         [nan, nan, nan,  ..., nan, nan, nan],
         [nan, nan, nan,  ..., nan, nan, nan],
         [nan, nan, nan,  ..., nan, nan, nan]],

        ...,

        [[nan, nan, nan,  ..., nan, nan, nan],
         [nan, nan, nan,  ..., nan, nan, nan],
         [nan, nan, nan,  ..., nan, nan, nan],
         ...,
         [nan, nan, nan,  ..., nan, nan, nan],
         [nan, nan, nan,  ..., nan, nan, nan],
         [nan, nan, nan,  ..., nan, nan, nan]],

        [[nan, nan, nan,  ..., nan, nan, nan],
         [nan, nan, nan,  ..., nan, nan, nan],
         [nan, nan, nan,  ..., nan, nan, nan],
         ...,
         [nan, nan, nan,  ..., nan, nan, nan],
         [nan, nan, nan,  ..., nan, nan, nan],
         [nan, nan, nan,  ..., nan, nan, nan]],

        [[nan, nan, nan,  ..., nan, nan, nan],
         [nan, nan, nan,  ..., nan, nan, nan],
         [nan, nan, nan,  ..., nan, nan, nan],
         ...,
         [nan, nan, nan,  ..., nan, nan, nan],
         [nan, nan, nan,  ..., nan, nan, nan],
         [nan, nan, nan,  ..., nan, nan, nan]]], device='cuda:0',
       grad_fn=<ViewBackward0>)

**Comment by Tai-Hsien** [comment]
Hello,

This only tells us that loss is nan. We might need to have more information. Could you check the y_pred before sending to the loss function?

Do you train from scratch? or conduct a transfer learning based on the pre-trained model we provide? Is this still the same 15-class problem? Does your training set also contain labels from 0 to 15 (0 is gingiva, UR7=label 1, UR6=label 2, and so on)?

**Comment by carina1754** [comment]
The output of the model is as follows. We have a total of 17 labels. Labeling (including wisdom teeth) was also performed from 0 to 16. Partially scanned data is also included in the training data, will this be a problem? We're working from scratch.
![image](https://user-images.githubusercontent.com/40535448/150903877-f6630d78-eaaa-424b-8c04-cfeb427dc34f.png)

     tensor([[[nan, nan, nan,  ..., nan, nan, nan],
     [nan, nan, nan,  ..., nan, nan, nan],
     [nan, nan, nan,  ..., nan, nan, nan],
     ...,
     [nan, nan, nan,  ..., nan, nan, nan],
     [nan, nan, nan,  ..., nan, nan, nan],
     [nan, nan, nan,  ..., nan, nan, nan]],

    [[nan, nan, nan,  ..., nan, nan, nan],
     [nan, nan, nan,  ..., nan, nan, nan],
     [nan, nan, nan,  ..., nan, nan, nan],
     ...,
     [nan, nan, nan,  ..., nan, nan, nan],
     [nan, nan, nan,  ..., nan, nan, nan],
     [nan, nan, nan,  ..., nan, nan, nan]],

    [[nan, nan, nan,  ..., nan, nan, nan],
     [nan, nan, nan,  ..., nan, nan, nan],
     [nan, nan, nan,  ..., nan, nan, nan],
     ...,
     [nan, nan, nan,  ..., nan, nan, nan],
     [nan, nan, nan,  ..., nan, nan, nan],
     [nan, nan, nan,  ..., nan, nan, nan]],

    ...,

    [[nan, nan, nan,  ..., nan, nan, nan],
     [nan, nan, nan,  ..., nan, nan, nan],
     [nan, nan, nan,  ..., nan, nan, nan],
     ...,
     [nan, nan, nan,  ..., nan, nan, nan],
     [nan, nan, nan,  ..., nan, nan, nan],
     [nan, nan, nan,  ..., nan, nan, nan]],

    [[nan, nan, nan,  ..., nan, nan, nan],
     [nan, nan, nan,  ..., nan, nan, nan],
     [nan, nan, nan,  ..., nan, nan, nan],
     ...,
     [nan, nan, nan,  ..., nan, nan, nan],
     [nan, nan, nan,  ..., nan, nan, nan],
     [nan, nan, nan,  ..., nan, nan, nan]],

    [[nan, nan, nan,  ..., nan, nan, nan],
     [nan, nan, nan,  ..., nan, nan, nan],
     [nan, nan, nan,  ..., nan, nan, nan],
     ...,
     [nan, nan, nan,  ..., nan, nan, nan],
     [nan, nan, nan,  ..., nan, nan, nan],
     [nan, nan, nan,  ..., nan, nan, nan]]], device='cuda:0',
   grad_fn=<ViewBackward0>)

     tensor([[[nan, nan, nan,  ..., nan, nan, nan],
     [nan, nan, nan,  ..., nan, nan, nan],
     [nan, nan, nan,  ..., nan, nan, nan],
     ...,
     [nan, nan, nan,  ..., nan, nan, nan],
     [nan, nan, nan,  ..., nan, nan, nan],
     [nan, nan, nan,  ..., nan, nan, nan]],

    [[nan, nan, nan,  ..., nan, nan, nan],
     [nan, nan, nan,  ..., nan, nan, nan],
     [nan, nan, nan,  ..., nan, nan, nan],
     ...,
     [nan, nan, nan,  ..., nan, nan, nan],
     [nan, nan, nan,  ..., nan, nan, nan],
     [nan, nan, nan,  ..., nan, nan, nan]],

    [[nan, nan, nan,  ..., nan, nan, nan],
     [nan, nan, nan,  ..., nan, nan, nan],
     [nan, nan, nan,  ..., nan, nan, nan],
     ...,
     [nan, nan, nan,  ..., nan, nan, nan],
     [nan, nan, nan,  ..., nan, nan, nan],
     [nan, nan, nan,  ..., nan, nan, nan]],

    ...,

    [[nan, nan, nan,  ..., nan, nan, nan],
     [nan, nan, nan,  ..., nan, nan, nan],
     [nan, nan, nan,  ..., nan, nan, nan],
     ...,
     [nan, nan, nan,  ..., nan, nan, nan],
     [nan, nan, nan,  ..., nan, nan, nan],
     [nan, nan, nan,  ..., nan, nan, nan]],

    [[nan, nan, nan,  ..., nan, nan, nan],
     [nan, nan, nan,  ..., nan, nan, nan],
     [nan, nan, nan,  ..., nan, nan, nan],
     ...,
     [nan, nan, nan,  ..., nan, nan, nan],
     [nan, nan, nan,  ..., nan, nan, nan],
     [nan, nan, nan,  ..., nan, nan, nan]],

    [[nan, nan, nan,  ..., nan, nan, nan],
     [nan, nan, nan,  ..., nan, nan, nan],
     [nan, nan, nan,  ..., nan, nan, nan],
     ...,
     [nan, nan, nan,  ..., nan, nan, nan],
     [nan, nan, nan,  ..., nan, nan, nan],
     [nan, nan, nan,  ..., nan, nan, nan]]], device='cuda:0',
   grad_fn=<ViewBackward0>)

**Comment by Tai-Hsien** [comment]
Interesting... Still no idea so far. I did performed a 17-class training before and didn't get any problems.

What does "partially scanned data" mean?

**Comment by carina1754** [comment]
![image](https://user-images.githubusercontent.com/40535448/150904430-63b6579d-b334-4f2d-a463-35f16fdcc15c.png)
I mean these data.

**Comment by carina1754** [comment]
If I send you 10 revisions of data, could you please test it? T^T

**Comment by Tai-Hsien** [comment]
Hello,

It will be a good idea to put the training dataset into the loss function without training. I feel the problem is from the calculation of loss function. If possible, also print out each variables line by line to see which one cause the nan.

**Comment by Tai-Hsien** [comment]
> If I send you 10 revisions of data, could you please test it? T^T

Sure, but probably not today. I can test it when I have some time.

**Comment by carina1754** [comment]
thank you. I'll let you know the results after I run the test.

**Comment by carina1754** [comment]
Can you give me your email?

**Comment by Tai-Hsien** [comment]
> Can you give me your email?

wu.5084@osu.edu

**Comment by carina1754** [comment]
> Hello,
> 
> It will be a good idea to put the training dataset into the loss function without training. I feel the problem is from the calculation of loss function. If possible, also print out each variables line by line to see which one cause the nan.

[Epoch: 1/200, Batch: 20/52] loss: 0.9449113249778748, dsc: 0.04990803264081478, sen: 0.18365011997520925, ppv: 0.08640087936073541

When I looked at the data, there was NaN data. When training the model, I checked it by outputting one by one. My mistake. Thank you so much for your help!

Thanks!
